### PR TITLE
avoid requiring score type to be escaped at compile-time

### PR DESCRIPTION
**Pull request by jfgrimm**

### CODE REVIEW

**Comment by rsharris** [comment]
@jfgrimm Why

**Comment by jfgrimm** [comment]
@rsharris on our systems, we build most software with compiler wrappers that inject rpath. The escaping being patched out in this PR is a bit of a nightmare to handle; we haven't run into any other software that does this kind of thing (out of >1000 unique software).
Since this should be an unobtrusive change, I thought I might as well make a PR for it

**Comment by rsharris** [comment]
OK thanks. At least now I know what problem you are trying to solve.

It would be helpful to know what system you had the problem on. I gather that is the only system you have tested your solution on. To accept the PR I would need to assess whether the change is supported on all the systems I expect people to use this package on.

I'm pretty old school and admittedly I have not heard of rpath before. Some googling tells me it is related to hard-coding a path in a binary (executable or library). But this repo does not use any libraries other than the standard C runtimes, and builds those into the executable rather than dynamically linking them (at least, that was/is my intent). So I don't understand why rpath injection (if that's the correct term) would be necessary, for *this* repo's build products. 

Another google hit tells me rpath is deprecated. But it looks like I'd have to dig a lot wider and deeper to fully understand all aspects of that. Skimming that info, I see hints of different behavior on different systems.

Examining your changes, it looks straightforward. However, I am somewhat confused as to what type the value of score_type has in your implementation. My concept of what is a valid logical clause in an `#if` did not include what you have (e.g. is D a string, or something more akin to a variable name), and so I am pondering whether I had learnt the concept incorrectly years ago or if this is something newer that might only be supported by more recent compilers.

So, to recap, I am not totally against the PR. But I need to fully understand all the reasons for it and what potential it has to break things on some systems.

**Comment by Flamefire** [comment]
The actual issue is that passing `-DFoo=\'I\'` through a compiler wrapper is hard, so it would be good to avoid that.

However the current solution does not work. I created a test (and possible solution) [on godbolt](https://godbolt.org/z/44oqfEMvv)

Basically: `#if score_type == I` does not compare the value of `score_type` to the character `I` but to a macro of that name. I'm not fully sure why but that always succeeds. I guess after macro substitution this becomes empty.   
So `scoreType` will always be `'I'` with this change which likely breaks the binary or at least produces unexpected results.

We can only compare integers and "character constants" with the preprocessor, but we [cannot](https://gcc.gnu.org/onlinedocs/gcc-4.5.1/cpp/Stringification.html#Stringification) create a character constant out of a macro value:
>     There is no way to convert a macro argument into a character constant.

What we can do is "rename" a macro by concatenating it with another one. Then we can use that to compare against "expected" names. I.e. we compare the value of the concatenated value to the value of another macro

The example above shows what happens with the current and the fixed version of the code when passing different values for `-Dscore_type` without quotes.   
For the current approach the first branch is always taken, I show this by issuing an error.

The other approach creates e.g. `SCORE_TYPE_F` out of a define which then expands to a character. A sanity check can error if the value is not valid. I.e. I propose:

```cpp
#define CONCAT(a, b) a ## b

#define SCORE_TYPE_I 'I'
#define SCORE_TYPE_F 'F'
#define SCORE_TYPE_D 'D'
// Indirection required to expand the macro argument
#define MAKE_SCORE_TYPE(c) CONCAT(SCORE_TYPE_, c)
#define scoreType MAKE_SCORE_TYPE(score_type)

#if scoreType != 'I' && scoreType != 'F' && scoreType != 'D'
#error "FAILED"
#endif
```

You can verify this on https://godbolt.org/z/Y1veqn1h1 it also shows that it correctly errors out on invalid values and the `-E` shows the final value of `scoreType`

**Comment by Flamefire** [comment]
Please have a look at https://github.com/lastz/lastz/pull/64 which implements the proposed solution
